File: server/scheduling.js
const db = require("./db");

/**
 * Generate available slots for a doctor on a given day.
 * Subtracts booked appointments and time-off blocks.
 */
function availableSlots(doctorId, dayStr, mode) {
  const dayDate = new Date(dayStr + "T00:00:00");
  const weekday = (dayDate.getDay() + 6) % 7; // JS: 0=Sun, convert to 0=Mon

  let rules = db.prepare(
    "SELECT * FROM availability_rules WHERE doctor_id = ? AND weekday = ?"
  ).all(doctorId, weekday);

  if (mode) {
    rules = rules.filter((r) => r.mode === mode);
  }

  const dayStart = dayStr + "T00:00:00";
  const dayEnd = dayStr + "T23:59:59";

  const booked = new Set(
    db.prepare(
      `SELECT starts_at FROM appointments
       WHERE doctor_id = ? AND starts_at >= ? AND starts_at <= ?
       AND status IN ('pending','confirmed')`
    ).all(doctorId, dayStart, dayEnd).map((a) => a.starts_at)
  );

  const offs = db.prepare(
    `SELECT starts_at, ends_at FROM time_off
     WHERE doctor_id = ? AND starts_at <= ? AND ends_at >= ?`
  ).all(doctorId, dayEnd, dayStart);

  const now = new Date().toISOString();
  const slots = [];

  for (const rule of rules) {
    const [sh, sm] = rule.start_time.split(":").map(Number);
    const [eh, em] = rule.end_time.split(":").map(Number);
    const startMin = sh * 60 + sm;
    const endMin = eh * 60 + em;
    const step = rule.slot_minutes;

    for (let m = startMin; m + step <= endMin; m += step) {
      const hh = String(Math.floor(m / 60)).padStart(2, "0");
      const mm = String(m % 60).padStart(2, "0");
      const slotStart = `${dayStr}T${hh}:${mm}:00`;

      const endM = m + step;
      const ehh = String(Math.floor(endM / 60)).padStart(2, "0");
      const emm = String(endM % 60).padStart(2, "0");
      const slotEnd = `${dayStr}T${ehh}:${emm}:00`;

      if (booked.has(slotStart)) continue;
      if (slotStart <= now) continue;

      const inOff = offs.some(
        (o) => o.starts_at < slotEnd && o.ends_at > slotStart
      );
      if (inOff) continue;

      slots.push({ starts_at: slotStart, ends_at: slotEnd, mode: rule.mode });
    }
  }

  slots.sort((a, b) => (a.starts_at < b.starts_at ? -1 : 1));
  return slots;
}

function slotMinutes(doctorId, weekday) {
  const rule = db.prepare(
    "SELECT slot_minutes FROM availability_rules WHERE doctor_id = ? AND weekday = ?"
  ).get(doctorId, weekday);
  return rule ? rule.slot_minutes : 30;
}

module.exports = { availableSlots, slotMinutes };

File: backend/app/services/scheduling.py
"""Slot generation: expand weekly availability rules into concrete open slots,
subtracting booked appointments and time-off."""
from datetime import datetime, timedelta, date, time

from sqlalchemy.orm import Session

from ..models import (
    Doctor, AvailabilityRule, Appointment, TimeOff, AppointmentStatus, VisitMode,
)


def _combine(d: date, t: time) -> datetime:
    return datetime.combine(d, t)


def available_slots(db: Session, doctor_id: str, day: date,
                    mode: VisitMode | None = None) -> list[dict]:
    rules = (
        db.query(AvailabilityRule)
        .filter(AvailabilityRule.doctor_id == doctor_id,
                AvailabilityRule.weekday == day.weekday())
        .all()
    )
    if mode:
        rules = [r for r in rules if r.mode == mode]

    day_start = _combine(day, time.min)
    day_end = day_start + timedelta(days=1)

    booked = {
        a.starts_at
        for a in db.query(Appointment).filter(
            Appointment.doctor_id == doctor_id,
            Appointment.starts_at >= day_start,
            Appointment.starts_at < day_end,
            Appointment.status.in_([AppointmentStatus.pending,
                                    AppointmentStatus.confirmed]),
        )
    }
    offs = db.query(TimeOff).filter(
        TimeOff.doctor_id == doctor_id,
        TimeOff.starts_at < day_end,
        TimeOff.ends_at > day_start,
    ).all()

    def in_time_off(s: datetime, e: datetime) -> bool:
        return any(o.starts_at < e and o.ends_at > s for o in offs)

    now = datetime.utcnow()
    slots: list[dict] = []
    for r in rules:
        cursor = _combine(day, r.start_time)
        end = _combine(day, r.end_time)
        step = timedelta(minutes=r.slot_minutes)
        while cursor + step <= end:
            slot_end = cursor + step
            if (cursor not in booked and cursor > now
                    and not in_time_off(cursor, slot_end)):
                slots.append({"starts_at": cursor, "ends_at": slot_end,
                              "mode": r.mode})
            cursor = slot_end
    slots.sort(key=lambda s: s["starts_at"])
    return slots
